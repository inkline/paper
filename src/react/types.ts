export type VNode = JSX.Element;

export interface ComponentContext {
    useSlot: (name: string) => void;
}

export type ConstructorType =
    | NumberConstructor
    | StringConstructor
    | BooleanConstructor
    | ObjectConstructor
    | ArrayConstructor;

export type ComponentProps<Props> = {
    [key in keyof Props]: ConstructorType | {
        type: ConstructorType;
        default?: Props[key] | (() => Props[key]);
    };
}

export interface ComponentDefinition<Props, State> {
    props?: ComponentProps<Props>;
    setup?(props: Props): State;
    render(state: Props & State, context: ComponentContext): VNode;
}
