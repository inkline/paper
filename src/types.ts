export type Ref<T> = { value: T };

export type UnwrapState<State> = {
    [key in keyof State]: (State[key] & Ref<State[key]>)['value']
};

export type ConstructorType =
    | NumberConstructor
    | StringConstructor
    | BooleanConstructor
    | ObjectConstructor
    | ArrayConstructor;

export type ComponentProps<Props> = {
    [key in keyof Props]: ConstructorType | {
        type?: ConstructorType;
        default?: Props[key] | (() => Props[key]);
    };
}

export interface SetupContext {
    emit: (eventName: string, ...args: any[]) => void
}

export interface RenderContext {
    slot(name?: string): any;
}

export interface ComponentDefinition<Props extends Record<string, any> = {}, State extends Record<string, any> = {}, VNode = any> {
    slots?: string[];
    emits?: string[];
    props?: ComponentProps<Props>;
    setup?(props: Props, context: SetupContext): State;
    render(state: Props & State, context: RenderContext): VNode;
}
