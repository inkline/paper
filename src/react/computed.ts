import { Ref } from '@inkline/ucd/types';

export function computed<T> (computeFn: () => T): Ref<T> {
    return {
        get value () {
            return computeFn();
        }
    };
}
