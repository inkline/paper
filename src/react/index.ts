export * from '@inkline/ucd/react/defineComponent';
export * from '@inkline/ucd/react/computed';
export * from '@inkline/ucd/react/h';
export * from '@inkline/ucd/react/helpers';
export * from '@inkline/ucd/react/ref';
export * from '@inkline/ucd/react/types';
