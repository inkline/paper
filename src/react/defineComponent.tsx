import { Slots, VNode } from '@inkline/ucd/react/types';
import { ComponentDefinition, ComponentProps, RenderContext, SetupContext } from '@inkline/ucd/types';
import {PropsWithChildren, useEffect, useRef, useState} from 'react';
import { capitalizeFirst } from '@inkline/ucd/helpers';
import { getSlotChildren, normalizeEventName } from '@inkline/ucd/react/helpers';

/**
 * Global providers, identified uniquely using a symbol or string
 */
const providers: {
    [key: string | symbol]: {
        state: any;
        setState(newValue: any): void;
        notify(newValue: any): void;
    };
} = {};

/**
 * Register provider
 *
 * @param identifier
 * @param value
 */
function registerProvider <T> (identifier: string | symbol, value: T): void {
    providers[identifier] = {
        state: value,
        setState (newValue: any) {
            providers[identifier].state = newValue;
        },
        notify () {}
    };
}

/**
 * Update provider
 *
 * @param identifier
 * @param value
 */
function updateProvider <T> (identifier: string | symbol, value: T): void {
    providers[identifier].setState(value);
    providers[identifier].notify(value);
}

/**
 * Define React component using Functional Component and named slots
 *
 * @param definition universal component definition
 */
export function defineComponent<Props extends Record<string, any> = {}, State extends Record<string, any> = {}> (
    definition: ComponentDefinition<Props, State, VNode>
) {
    const slots: Slots = {};

    const Component: {
        (props: PropsWithChildren<Props>, context?: any): VNode;
        [key: string]: any;
    } = (props) => {
        /**
         * Setup context
         */
        const setupContext: SetupContext = {
            /**
             * Emit event
             *
             * @param eventName
             * @param args
             */
            emit: (eventName, ...args) => {
                props[normalizeEventName(eventName)]?.(...args);
            },
            /**
             * Provide value to consumers
             *
             * @param identifier
             * @param value
             */
            provide: (identifier, value) => {
                useEffect(() => {
                    if (providers[identifier]) {
                        updateProvider(identifier, value);
                    } else {
                        registerProvider(identifier, value);
                    }
                });
            },
            /**
             * Inject value from providers
             *
             * @param identifier
             * @param defaultValue
             */
            inject: (identifier, defaultValue) => {
                const [, setState] = useState();

                defaultValue = typeof defaultValue === 'function' ? (defaultValue as () => any)() : defaultValue;

                if (!providers[identifier]) {
                    registerProvider(identifier, defaultValue);
                }

                providers[identifier].notify = setState;
                return providers[identifier].state;
            }
        };

        /**
         * Render context
         */
        const renderContext: RenderContext = {
            slot (name: string = 'default') {
                const children = Array.isArray(props.children) ? props.children : [props.children];

                return getSlotChildren(name, slots, children);
            }
        };

        /**
         * State and props
         */
        const state = definition.setup
            ? { ...props, ...definition.setup(props, setupContext) }
            : props as Props & State;

        /**
         * Render
         */
        return definition.render(state, renderContext);
    };

    /**
     * Slots
     */
    ['default'].concat(definition.slots || []).forEach((name) => {
        slots[name] = () => null;
        slots[name].key = name;
        Component[capitalizeFirst(name)] = slots[name];
    });

    /**
     * Default props
     */
    if (definition.props) {
        Component.defaultProps = Object.keys(definition.props)
            .reduce((acc, propName) => {
                const propType = (definition.props as ComponentProps<any>)[propName];

                if (typeof propType === 'object' && propType.default) {
                    acc[propName] = typeof propType.default === 'function'
                        ? propType.default()
                        : propType.default;
                }

                return acc;
            }, {} as Record<string, any>);
    }

    return Component;
}
