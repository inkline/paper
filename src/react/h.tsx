import React from 'react';
import { VNode } from '@inkline/paper/react/types';

export function h (
    type: string,
    props?: any,
    children?: (string | number | boolean | VNode)[]
): VNode {
    return React.createElement(type, props, children) as VNode;
}
