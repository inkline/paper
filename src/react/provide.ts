import { useEffect, useState } from 'react';

/**
 * Global providers, identified uniquely using a symbol or string
 */
const providers: {
    [key: string | symbol]: {
        state: any;
        setState(newValue: any): void;
        notify(newValue: any): void;
    };
} = {};

/**
 * Register provider
 *
 * @param identifier
 * @param value
 */
function registerProvider <T> (identifier: string | symbol, value: T): void {
    providers[identifier] = {
        state: value,
        setState (newValue: any) {
            providers[identifier].state = newValue;
        },
        notify () {}
    };
}

/**
 * Update provider
 *
 * @param identifier
 * @param value
 */
function updateProvider <T> (identifier: string | symbol, value: T): void {
    providers[identifier].setState(value);
    providers[identifier].notify(value);
}

/**
 * Provide value to consumers
 *
 * @param identifier
 * @param value
 * @param dependencies
 */
export const provide = <T>(identifier: string | symbol, value: T, dependencies: any[] = []): void => {
    useEffect(() => {
        if (providers[identifier]) {
            updateProvider(identifier, value);
        } else {
            registerProvider(identifier, value);
        }
    }, dependencies);
};

/**
 * Inject value from providers
 *
 * @param identifier
 * @param defaultValue
 */
export const inject = <T>(identifier: string | symbol, defaultValue?: T | (() => T)): T | undefined => {
    const [, setState] = useState();

    if (!providers[identifier]) {
        defaultValue = typeof defaultValue === 'function'
            ? (defaultValue as () => any)()
            : defaultValue;

        registerProvider(identifier, defaultValue);
    }

    providers[identifier].notify = setState;
    return providers[identifier].state;
};
