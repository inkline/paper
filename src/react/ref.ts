import { useState } from 'react';
import { Ref } from '@inkline/ucd/types';

export function ref<T> (initialValue: T): Ref<T> {
    const [state, setState] = useState<T>(initialValue);

    return {
        get value () {
            return state;
        },
        set value (value: T) {
            setState(value);
        }
    };
}
