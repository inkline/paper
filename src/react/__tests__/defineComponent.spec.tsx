import React from 'react';
import {
    defineComponent,
    h,
    ref,
    Events
} from '@inkline/ucd/react';
import { fireEvent, render } from '@testing-library/react';
import { Ref } from '@inkline/ucd/types';

describe('react', () => {
    describe('defineComponent()', () => {
        it('should render component', () => {
            const Component = defineComponent({
                render () {
                    return h('div');
                }
            });

            const wrapper = render(<Component /> as any);
            expect(wrapper.container.firstChild).toMatchSnapshot();
        });

        it('should render component with state', () => {
            const Component = defineComponent<{}, { count: Ref<number>; }>({
                setup () {
                    const count = ref(0);

                    return { count };
                },
                render (state) {
                    return h('button', {}, [
                        `${state.count.value}`
                    ]);
                }
            });

            const wrapper = render(<Component /> as any);
            expect(wrapper.container.firstChild).toMatchSnapshot();
        });

        it('should render component with state and props', () => {
            const Component = defineComponent<{ initialValue: number; }, { count: Ref<number>; }>({
                setup (props) {
                    const count = ref(props.initialValue);

                    return { count };
                },
                render (state) {
                    return h('button', {}, [
                        `${state.count.value}`
                    ]);
                }
            });

            const wrapper = render(<Component initialValue={1} /> as any);
            expect(wrapper.container.firstChild).toMatchSnapshot();
        });

        it('should render component with state, props, and onClick state function', () => {
            const Component = defineComponent<{ initialValue: number; }, { count: Ref<number>; onClick:() => void }>({
                setup (props) {
                    const count = ref(props.initialValue);

                    const onClick = () => {
                        count.value += 1;
                    };

                    return { count, onClick };
                },
                render (state) {
                    return h('button', { onClick: state.onClick }, [
                        `${state.count.value}`
                    ]);
                }
            });

            const wrapper = render(<Component initialValue={1} /> as any);
            fireEvent.click(wrapper.container.firstChild as Element);
            expect(wrapper.container.firstChild).toMatchSnapshot();
        });

        describe('context', () => {
            describe('slot()', () => {
                it('should render react component with default slot, one child', () => {
                    const Component = defineComponent<{}, {}>({
                        render (state, ctx) {
                            return h('div', {}, [
                                ctx.slot()
                            ]);
                        }
                    });

                    const wrapper = render(<Component>
                        <span>Child</span>
                    </Component> as any);
                    expect(wrapper.container.firstChild).toMatchSnapshot();
                });

                it('should render react component with default slot, multiple children', () => {
                    const Component = defineComponent<{}, {}>({
                        render (state, ctx) {
                            return h('div', {}, [
                                ctx.slot()
                            ]);
                        }
                    });

                    const wrapper = render(<Component>
                        <span>Child 1</span>
                        <span>Child 2</span>
                        <span>Child 3</span>
                    </Component> as any);
                    expect(wrapper.container.firstChild).toMatchSnapshot();
                });

                it('should render react component with named default slot', () => {
                    const Component = defineComponent<{}, {}>({
                        render (state, ctx) {
                            return h('div', {}, [
                                ctx.slot()
                            ]);
                        }
                    });

                    const wrapper = render(<Component>
                        <Component.Default>
                            <span>Child</span>
                        </Component.Default>
                    </Component> as any);
                    expect(wrapper.container.firstChild).toMatchSnapshot();
                });

                it('should render react component with aggregated default slot and named default slot', () => {
                    const Component = defineComponent<{}, {}>({
                        render (state, ctx) {
                            return h('div', {}, [
                                ctx.slot()
                            ]);
                        }
                    });

                    const wrapper = render(<Component>
                        <span>Default Slot Child 1</span>
                        <Component.Default>
                            <span>Named Default Slot Child</span>
                        </Component.Default>
                        <span>Default Slot Child 2</span>
                    </Component> as any);
                    expect(wrapper.container.firstChild).toMatchSnapshot();
                });

                it('should render react component with named slots', () => {
                    const Component = defineComponent<{}, {}>({
                        slots: [
                            'header',
                            'footer'
                        ],
                        emits: [
                            'onUpdateModelValue'
                        ],
                        render (state, ctx) {
                            return h('div', {}, [
                                ctx.slot('header'),
                                ctx.slot(),
                                ctx.slot('footer')
                            ]);
                        }
                    });

                    const wrapper = render(<Component>
                        <Component.Header>
                            <span>Header</span>
                        </Component.Header>
                        <span>Body</span>
                        <Component.Footer>
                            <span>Footer</span>
                        </Component.Footer>
                    </Component> as any);
                    expect(wrapper.container.firstChild).toMatchSnapshot();
                });
            });

            describe('emit()', () => {
                it('should emit event by attaching on[EventName] callback', () => {
                    const Component = defineComponent<{ [key: string]: any; }, { onClick(): void }>({
                        emits: [
                            'click'
                        ],
                        setup (props, ctx) {
                            const onClick = () => {
                                ctx.emit('click');
                            };

                            return { onClick };
                        },
                        render (state) {
                            return h('button', { onClick: state.onClick }, ['Button']);
                        }
                    });

                    const onClick = vi.fn();
                    const wrapper = render(<Component onClick={onClick} /> as any);
                    fireEvent.click(wrapper.container.firstChild as Element);
                    expect(onClick).toHaveBeenCalled();
                });

                it('should emit event with arguments', () => {
                    const Component = defineComponent<{ [key: string]: any; }, { onClick(event: Event): void }>({
                        emits: [
                            'click'
                        ],
                        setup (props, ctx) {
                            const onClick = (event: Event) => {
                                ctx.emit('click', event);
                            };

                            return { onClick };
                        },
                        render (state) {
                            return h('button', { onClick: state.onClick }, ['Button']);
                        }
                    });

                    const onClick = vi.fn();
                    const wrapper = render(<Component onClick={onClick} /> as any);
                    fireEvent.click(wrapper.container.firstChild as Element);
                    expect(onClick).toHaveBeenCalled();
                    expect(onClick.mock.calls[0][0]).toEqual(expect.any(Object));
                });

                it('should render component with modelValue and update:modelValue', () => {
                    const Component = defineComponent<{ modelValue: number; [key: string]: any; }, { onClick(): void }>({
                        emits: [
                            'update:modelValue'
                        ],
                        setup (props, ctx) {
                            const onClick = () => {
                                ctx.emit('update:modelValue', props.modelValue + 1);
                            };

                            return { onClick };
                        },
                        render (state) {
                            return h('button', { onClick: state.onClick }, ['Button']);
                        }
                    });

                    let value = 3;
                    const onUpdate = vi.fn((newValue) => { value = newValue; });
                    const wrapper = render(<Component modelValue={value} onUpdate:modelValue={onUpdate} /> as any);

                    fireEvent.click(wrapper.container.firstChild as Element);

                    expect(onUpdate).toHaveBeenCalled();
                    expect(value).toEqual(4);
                });

                it('should render component with input field', () => {
                    const Component = defineComponent<{ modelValue: string; [key: string]: any; }, { onChange(event: Event): void }>({
                        props: {
                            modelValue: {
                                type: String,
                                default: ''
                            }
                        },
                        emits: [
                            'update:modelValue'
                        ],
                        setup (props, ctx) {
                            const onChange = (event: Event) => {
                                ctx.emit('update:modelValue', (event.target as HTMLInputElement).value);
                            };

                            return { onChange };
                        },
                        render (state) {
                            return h('input', { value: state.modelValue, [Events.onInputChange]: state.onChange });
                        }
                    });

                    let value = '';
                    const onUpdate = vi.fn((newValue) => { value = newValue; });
                    const wrapper = render(<Component modelValue={value} onUpdate:modelValue={onUpdate} /> as any);

                    fireEvent.change(wrapper.container.firstChild as Element, { target: { value: 'abc' } });

                    expect(onUpdate).toHaveBeenCalled();
                    expect(value).toEqual('abc');
                });
            });

            describe('provide/inject()', () => {
                it('should provide data to children', async () => {
                    const identifier = Symbol('provide-reactive');
                    const Provider = defineComponent<{}, {}>({
                        setup (props, ctx) {
                            ctx.provide(identifier, 'value');

                            return {};
                        },
                        render (state, ctx) {
                            return h('div', {}, [
                                ctx.slot()
                            ]);
                        }
                    });

                    const Consumer = defineComponent<{}, { providedValue: string; }>({
                        setup (props, ctx) {
                            const providedValue = ctx.inject<string>(identifier);

                            return { providedValue };
                        },
                        render (state) {
                            return h('div', {}, [
                                state.providedValue
                            ]);
                        }
                    });

                    const wrapper = render(<Provider>
                        <Consumer />
                    </Provider> as any);
                    expect(wrapper.container.firstChild).toMatchSnapshot();
                });

                it.only('should provide reactive data to children', async () => {
                    const identifier = Symbol('provide-reactive');
                    const Provider = defineComponent<{}, { onClick(): void }>({
                        setup (props, ctx) {
                            const count = ref(0);
                            const onClick = () => { count.value += 1; };

                            ctx.provide(identifier, count);

                            return { onClick };
                        },
                        render (state, ctx) {
                            return h('button', { onClick: state.onClick }, [
                                ctx.slot()
                            ]);
                        }
                    });

                    const Consumer = defineComponent<{}, { providedValue: Ref<number>; }>({
                        setup (props, ctx) {
                            const providedValue = ctx.inject<Ref<number>>(identifier);

                            return { providedValue };
                        },
                        render (state) {
                            return h('div', {}, [
                                state.providedValue?.value
                            ]);
                        }
                    });

                    const wrapper = render(<Provider>
                        <Consumer />
                    </Provider> as any);
                    fireEvent.click(wrapper.container.firstChild as Element);
                    expect(await wrapper.findByText('1')).toBeTruthy();
                    expect(wrapper.container.firstChild).toMatchSnapshot();
                });

                it('should provide reactive data to children 2', async () => {
                    const identifier = Symbol('provide-reactive');

                    const Provider1 = defineComponent<{}, { onClick(): void }>({
                        setup (props, ctx) {
                            const count = ref(0);
                            const onClick = () => { count.value += 1; };

                            ctx.provide(identifier, count);

                            return { onClick };
                        },
                        render (state, ctx) {
                            return h('button', { onClick: state.onClick }, [
                                ctx.slot()
                            ]);
                        }
                    });

                    const Provider2 = defineComponent<{}, { onClick(): void }>({
                        setup (props, ctx) {
                            const count = ref(0);
                            const onClick = () => { count.value += 1; };

                            ctx.provide(identifier, count);

                            return { onClick };
                        },
                        render (state, ctx) {
                            return h('button', { onClick: state.onClick }, [
                                ctx.slot()
                            ]);
                        }
                    });

                    const Consumer = defineComponent<{}, { providedValue: Ref<number>; }>({
                        setup (props, ctx) {
                            const providedValue = ctx.inject<Ref<number>>(identifier);

                            return { providedValue };
                        },
                        render (state) {
                            return h('div', {}, [
                                state.providedValue?.value
                            ]);
                        }
                    });

                    const wrapper = render(<Provider1>
                        <Consumer />
                    </Provider1> as any);
                    fireEvent.click(wrapper.container.firstChild as Element);
                    expect(wrapper.findByText('1')).toBeTruthy();

                    const wrapper2 = render(<Provider2>
                        <Consumer />
                    </Provider2> as any);
                    expect(await wrapper2.findByText('1')).not.toBeTruthy();
                });
            });
        });
    });
});
