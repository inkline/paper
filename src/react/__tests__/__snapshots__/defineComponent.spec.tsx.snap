// Vitest Snapshot v1

exports[`react > defineComponent() > context > provide/inject() > should provide reactive data to children 1`] = `
<button>
  <div>
    1
  </div>
</button>
`;
